web/src/engine/predictive-text/worker-thread/src/main/transform-subsets.ts

```diff
@@ -0,0 +1,144 @@
+   * for all Transforms in this subset.
+   */
+  insert: string
+}
+
+/**
+ * Places all Transforms in the distribution into distinct subsets where BOTH
+ * the following properties match for all entries in the same set:
+ * 1. `deleteLeft`
+ * 2. `insert.length`
+ *
+ * The subset instances are keyed by just `insert.length`, despite both serving
+ * as criteria for building the subsets.
+ *
+ * It also computes the sum total probability mass represented by the set.
+ *
+ * Array index: the deleteLeft value (which is always non-negative) Inner map
+ * index:  the length of the insert string.
+ * @param distribution
+ */
+export function subsetByInterval(
+  distribution: Distribution<Transform>
+): TransformSubset<number>[][] {
+  const returnRoot: TransformSubset<number>[][] = [];
+
+  distribution.forEach((entry) => {
+    const { sample, p } = entry;
+
+    const dl = sample.deleteLeft;
+    const deleteBin = returnRoot[dl] ?? [];
+
+    const ins = KMWString.length(sample.insert);
+    const subset: TransformSubset<number> = deleteBin[ins] ?? { key: ins, insert: "", entries: [], cumulativeMass: 0 };
+
+    subset.entries.push(entry);
+    subset.cumulativeMass += p;
+
+    deleteBin[ins] = subset;
+    returnRoot[dl] = deleteBin;
+  });
+
+  return returnRoot;
+}
+
+/**
+ * Places all Transforms in the distribution into distinct subsets where the
+ * character at the specified index within their Transform matches.
+ *
+ * It also computes the sum total probability mass represented by the set.
+ *
+ * Map key:  the character shared (at the specified subindex) by each Transform
+ * in the subset
+ * @param intervalSubset
+ * @param subIndex
+ */
+export function subsetByChar(intervalSubset: TransformSubset<number>, subIndex: number, toKey?: (s: string) => string): Map<string, TransformSubset<string>> {
+  toKey = toKey ?? (x => x);
+  const returnRoot: Map<string, TransformSubset<string>> = new Map();
+
+  // If there the input catalog doesn't support the subindex, just return everything in
+  // one bulk catalog.
+  if(intervalSubset.key <= subIndex) {
+    const everything: TransformSubset<string> = {...intervalSubset, key: ''};
+    returnRoot.set('', everything);
+    return returnRoot;
+  }
+
+  intervalSubset.entries.forEach((entry) => {
+    const { sample, p } = entry;
+    const char = toKey(KMWString.charAt(sample.insert, subIndex));
+
+    const subset: TransformSubset<string> = returnRoot.get(char) ?? { key: char, insert: intervalSubset.insert, entries: [], cumulativeMass: 0 };
+
+    subset.entries.push(entry);
+    subset.cumulativeMass += p;
+
+    returnRoot.set(char, subset);
+  });
+
+  return returnRoot;
+}
+
+/**
+ * Merges two or more TransformCatalog instances into a single instance keyed by
+ * the value passed in as a parameter.
+ * @param subsets
+ * @param key
+ * @returns
+ */
+export function mergeSubset<T, K>(subsets: TransformSubset<T>[], key: K): TransformSubset<K> {
```

Comment:
Doesn't seem to be used or tested.

Reply:
See https://github.com/keymanapp/keyman/pull/14592#discussion_r2301113563.